File: EMS/Network/EMSSlotsParser.m
//
//  EMSSlotsParser.m
//

#import "EMS-Swift.h"

#import "EMSSlotsParser.h"
#import "EMSSlot.h"

#import "CJCollection.h"
#import "CJItem.h"

#import "EMSDateConverter.h"

static const DDLogLevel ddLogLevel = DDLogLevelDebug;

@implementation EMSSlotsParser

- (NSArray *)processData:(NSData *)data forHref:(NSURL *)href error:(NSError **)error {
    NSError *parseError = nil;

    CJCollection *collection = [CJCollection collectionForNSData:data error:&parseError];

    if (!collection) {
        DDLogError(@"Failed to retrieve slots %@ - %@ - %@", href, parseError, [parseError userInfo]);

        *error = parseError;

        return [NSArray array];
    }

    NSMutableArray *temp = [[NSMutableArray alloc] init];

    [collection.items enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
        CJItem *item = (CJItem *) obj;

        EMSSlot *slot = [[EMSSlot alloc] init];

        slot.href = item.href;

        [item.data enumerateObjectsUsingBlock:^(id dataObj, NSUInteger dataIdx, BOOL *dataStop) {
            NSDictionary *dict = (NSDictionary *) dataObj;

            NSString *field = dict[@"name"];
            NSString *value = dict[@"value"];

            if ([@"start" isEqualToString:field]) {
                slot.start = [EMSDateConverter dateFromString:value];
            }
            if ([@"duration" isEqualToString:field]) {
                slot.duration = [value integerValue];
            }

        }];

        [temp addObject:slot];
    }];

    return [NSArray arrayWithArray:temp];
}

- (void)parseData:(NSData *)data forHref:(NSURL *)href {
    NSError *error;

    NSArray *collection = [self processData:data forHref:href error:&error];

    [self.delegate finishedSlots:collection forHref:href error:error];
}

@end

File: EMS/Network/Parsers/EMSSlotsParser.swift
import Foundation
import CocoaLumberjack

@objc protocol EMSSlotsParserDelegate {
    optional func finishedSlots(slots: [EMSSlot], forHref href:NSURL, error:NSError?)
}

@objc class EMSSlotsParser : NSObject {
    var delegate : EMSSlotsParserDelegate?
    
    func processData(data: NSData, forHref href:NSURL) throws -> [EMSSlot] {
        
        do {
            let collection = try CJCollection(forNSData:data)
            
            return collection.items.map({ item in
                let slot = EMSSlot()
                
                slot.href = item.href
                
                for datum in item.data {
                    if let field = datum["name"] as? String {
                        if let value = datum["value"] as? String {
                            switch (field) {
                            case "start":
                                slot.start = EMSDateConverter.dateFromString(value)
                            default:
                                DDLogDebug("Ignored string field \(field) for slot")
                            }
                        }
                    
                        if let value = datum["value"] as? Int {
                            switch (field) {
                            case "duration":
                                slot.slotDuration = value
                            default:
                                DDLogDebug("Ignored integer field \(field) for slot")
                            }
                        }
                    }
                }

                return slot
            })
        } catch let error as NSError {
            DDLogError("Failed to retrieve slots \(href) - \(error)")
            
            throw error
        }
    }
    
    func parseData(data: NSData, forHref url: NSURL) {
        DDLogDebug("Hofbnveopbne49+bn49+ubn+e9")
        do {
            let collection = try self.processData(data, forHref: url)
            
            self.delegate?.finishedSlots?(collection, forHref: url, error: nil)
        } catch let error as NSError {
            self.delegate?.finishedSlots?([], forHref: url, error: error)
        }
    }
}
